File: libs/language/src/lib/utils/BlueIdToCid.ts
import { CID } from 'multiformats/cid';
import * as raw from 'multiformats/codecs/raw';
import { sha256 } from 'multiformats/hashes/sha2';
import { Base58 } from './Base58';
import { create } from 'multiformats/hashes/digest';

export class BlueIdToCid {
  static convert(blueId: string): string {
    const sha256Bytes: Uint8Array = Base58.decode(blueId);
    const multihash = create(sha256.code, sha256Bytes);
    const cid = CID.create(1, raw.code, multihash);

    return cid.toString();
  }
}

File: references/java-sdk/src/main/java/blue/language/provider/ipfs/BlueIdToCid.java
package blue.language.provider.ipfs;

import blue.language.utils.Base58;
import org.apache.commons.codec.binary.Base32;

public class BlueIdToCid {

    public static String convert(String blueId) {
        byte[] sha256Bytes = Base58.decode(blueId);

        // Create the multihash bytes for SHA-256 (0x12 for the hash function and 0x20 for the length)
        byte[] multihash = new byte[2 + sha256Bytes.length];
        multihash[0] = 0x12; // SHA-256
        multihash[1] = 0x20; // 32 bytes (256 bits)
        System.arraycopy(sha256Bytes, 0, multihash, 2, sha256Bytes.length);

        // Create the CIDv1 bytes with version byte (0x01) and codec for raw (0x55)
        byte[] cidBytes = new byte[2 + multihash.length];
        cidBytes[0] = 0x01; // CIDv1
        cidBytes[1] = 0x55; // raw binary data
        System.arraycopy(multihash, 0, cidBytes, 2, multihash.length);

        // Encode the CIDv1 with Base32
        Base32 base32 = new Base32();
        String cid = "b" + base32.encodeAsString(cidBytes).toLowerCase().replaceAll("=", "");

        return cid;
    }

}